File: apps/macos/Lumina/Services/SidecarReadiness.swift
import Foundation

enum SidecarEngineStatus: Equatable {
    case starting
    case running
    case stopped
    case failed
}

/// Outcome of one health probe while waiting for a freshly spawned sidecar.
enum HealthPollDecision: Equatable {
    case ready
    case keepWaiting
    /// `/health` answered but chunker/core identity does not match this app.
    case incompatible
    /// Owned process exited before becoming healthy.
    case processExited
}

/// Pure readiness decisions for SidecarManager (unit-testable).
enum SidecarReadiness {
    /// Must match lumina-core `CHUNKER_VERSION`. A mismatch never marks the sidecar ready,
    /// so library / news / settings all refuse to load.
    static let expectedChunkerVersion = "16"

    /// GET /health and POST /shutdown must not wait on a wedged event loop.
    static let probeTimeoutSeconds: TimeInterval = 2

    /// Extreme upper bound for one launch attempt's health polling (~30s of delays).
    static let healthPollBudgetSeconds: TimeInterval = 30

    static let messageTimeout = "AI 引擎启动超时，请重试或退出。"
    static let messageProcessExited = "AI 引擎进程已退出，请重试。"
    static let messageIncompatible =
        "AI 引擎版本与应用不匹配，请更新应用或重启引擎。"

    static func engineStatus(
        isRunning: Bool,
        isBootstrapping: Bool,
        userStopped: Bool,
        launchError: String?
    ) -> SidecarEngineStatus {
        if isBootstrapping { return .starting }
        if isRunning { return .running }
        if userStopped { return .stopped }
        if launchError != nil { return .failed }
        return .stopped
    }

    static func statusLabel(_ status: SidecarEngineStatus) -> String {
        switch status {
        case .starting: return "正在启动…"
        case .running: return "运行中"
        case .stopped: return "已停止"
        case .failed: return "启动失败"
        }
    }

    static func shouldAutoStart(userStopped: Bool) -> Bool {
        !userStopped
    }

    /// Quit / Settings 停止 must kill whoever holds the port, not only this session's Process.
    static func mustKillPortListenerOnStop(hasOwnedProcess: Bool, portOccupied: Bool) -> Bool {
        hasOwnedProcess || portOccupied
    }

    static func shouldKillListenerBeforeLaunch(
        healthResponded: Bool,
        shouldReplaceOrphan: Bool,
        portOccupied: Bool
    ) -> Bool {
        if healthResponded { return shouldReplaceOrphan }
        return portOccupied
    }

    static func shouldReuseLeftover(healthResponded: Bool, shouldReplaceOrphan: Bool) -> Bool {
        healthResponded && !shouldReplaceOrphan
    }

    /// Align with Windows: `1.1.0.0` and `1.1.0` are the same marketing version.
    static func normalizeVersion(_ version: String?) -> String {
        guard let version, !version.isEmpty else { return "" }
        let parts = version.split(separator: ".")
        if parts.count >= 3 {
            return "\(parts[0]).\(parts[1]).\(parts[2])"
        }
        return version
    }

    static func isCompatible(
        chunkerVersion: String?,
        coreVersion: String?,
        expectedCoreVersion: String
    ) -> Bool {
        !expectedCoreVersion.isEmpty
            && chunkerVersion == expectedChunkerVersion
            && normalizeVersion(coreVersion) == normalizeVersion(expectedCoreVersion)
    }

    /// Decide whether to keep polling, succeed, or fail-fast for this launch attempt.
    ///
    /// A responding but incompatible health must not burn the full 30s poll budget.
    static func evaluateHealthPoll(
        healthResponded: Bool,
        compatible: Bool,
        processStillRunning: Bool?
    ) -> HealthPollDecision {
        if healthResponded {
            return compatible ? .ready : .incompatible
        }
        if processStillRunning == false {
            return .processExited
        }
        return .keepWaiting
    }

    /// Delay *after* probe `afterProbeIndex` (0 = first probe was immediate).
    /// Fast early probes, then settle at 250ms; total budget remains ~30s.
    static func healthPollDelayNanoseconds(afterProbeIndex: Int) -> UInt64 {
        if afterProbeIndex < 20 { return 50_000_000 }
        if afterProbeIndex < 40 { return 100_000_000 }
        return 250_000_000
    }

    static func launchFailureMessage(for decision: HealthPollDecision) -> String {
        switch decision {
        case .incompatible:
            return messageIncompatible
        case .processExited:
            return messageProcessExited
        case .ready, .keepWaiting:
            return messageTimeout
        }
    }

    static func incompatibleDetailMessage(
        chunkerVersion: String?,
        coreVersion: String?,
        expectedCoreVersion: String
    ) -> String {
        let engine = "\(normalizeVersion(coreVersion))/chunker \(chunkerVersion ?? "?")"
        let expected = "\(normalizeVersion(expectedCoreVersion))/chunker \(expectedChunkerVersion)"
        return "AI 引擎版本与应用不匹配（引擎 \(engine)，应用期望 \(expected)）。请更新应用或重启引擎。"
    }

    /// Whether a process already bound to the sidecar port must be killed before launch.
    ///
    /// Chunker version alone is not enough: ingest/decode fixes ship under the same
    /// `CHUNKER_VERSION`. A leftover frozen binary must yield to this app's version
    /// and to a newer bundled executable (in-place rebuild of the same version).
    static func shouldReplaceOrphan(
        chunkerVersion: String?,
        coreVersion: String?,
        expectedCoreVersion: String,
        hasBundledSidecar: Bool,
        orphanExecutable: String?,
        bundledExecutable: String?,
        orphanStartedAt: Date?,
        bundledModifiedAt: Date?
    ) -> Bool {
        if !isCompatible(
            chunkerVersion: chunkerVersion,
            coreVersion: coreVersion,
            expectedCoreVersion: expectedCoreVersion
        ) {
            return true
        }
        if !hasBundledSidecar {
            return looksLikeFrozenSidecar(orphanExecutable)
        }
        if !pathsReferToSameFile(orphanExecutable, bundledExecutable) {
            return true
        }
        guard let started = orphanStartedAt, let modified = bundledModifiedAt else {
            return true
        }
        return started < modified
    }

    static func looksLikeFrozenSidecar(_ path: String?) -> Bool {
        guard let path, !path.isEmpty else { return false }
        let normalized = path.replacingOccurrences(of: "\\", with: "/")
        return normalized.contains("/Contents/Resources/lumina-core/")
    }

    static func pathsReferToSameFile(_ left: String?, _ right: String?) -> Bool {
        guard let left, let right, !left.isEmpty, !right.isEmpty else { return false }
        return URL(fileURLWithPath: left).standardizedFileURL.path
            == URL(fileURLWithPath: right).standardizedFileURL.path
    }

    /// After polling for bootstrap to begin, should we invoke ensureRunning ourselves?
    static func shouldInvokeEnsureRunning(
        isRunning: Bool,
        isBootstrapping: Bool,
        launchError: String?,
        sawBootstrapStart: Bool
    ) -> Bool {
        if isRunning || isBootstrapping || launchError != nil {
            return false
        }
        return !sawBootstrapStart
    }

    static func isReady(isRunning: Bool, launchError: String?) -> Bool {
        isRunning && launchError == nil
    }
}

File: apps/windows/Lumina/Services/SidecarReadiness.cs
namespace Lumina.Services;

public sealed class SidecarHealth
{
    public string Status { get; set; } = "";
    public int? Pid { get; set; }
    public string? ChunkerVersion { get; set; }
    public string? CoreVersion { get; set; }
    public string? Executable { get; set; }
    public long? StartedAt { get; set; }
    public int? UptimeMs { get; set; }
}

public enum HealthPollDecision
{
    Ready,
    KeepWaiting,
    /// `/health` answered but chunker/core identity does not match this app.
    Incompatible,
    /// Owned process exited before becoming healthy.
    ProcessExited,
}

/// Decisions for replacing a leftover lumina-core process on the fixed sidecar port.
public static class SidecarReadiness
{
    public const string ExpectedChunkerVersion = "16";
    public const double HealthPollBudgetSeconds = 30;
    public const string MessageTimeout = "AI 引擎启动超时，请重试或退出。";
    public const string MessageProcessExited = "AI 引擎进程已退出，请重试。";
    public const string MessageIncompatible = "AI 引擎版本与应用不匹配，请更新应用或重启引擎。";

    public static string NormalizeVersion(string? version)
    {
        if (string.IsNullOrWhiteSpace(version)) return "";
        var parts = version.Split('.');
        return parts.Length >= 3 ? string.Join(".", parts[0], parts[1], parts[2]) : version;
    }

    public static bool IsCompatible(string? chunkerVersion, string? coreVersion, string expectedCoreVersion)
    {
        return !string.IsNullOrEmpty(expectedCoreVersion)
            && chunkerVersion == ExpectedChunkerVersion
            && NormalizeVersion(coreVersion) == NormalizeVersion(expectedCoreVersion);
    }

    /// Decide whether to keep polling, succeed, or fail-fast for this launch attempt.
    /// A responding but incompatible health must not burn the full 30s poll budget.
    public static HealthPollDecision EvaluateHealthPoll(
        bool healthResponded,
        bool compatible,
        bool? processStillRunning)
    {
        if (healthResponded)
            return compatible ? HealthPollDecision.Ready : HealthPollDecision.Incompatible;
        if (processStillRunning == false)
            return HealthPollDecision.ProcessExited;
        return HealthPollDecision.KeepWaiting;
    }

    /// Delay after probe `afterProbeIndex` (0 = first probe was immediate).
    public static int HealthPollDelayMilliseconds(int afterProbeIndex)
    {
        if (afterProbeIndex < 20) return 50;
        if (afterProbeIndex < 40) return 100;
        return 250;
    }

    public static string LaunchFailureMessage(HealthPollDecision decision) => decision switch
    {
        HealthPollDecision.Incompatible => MessageIncompatible,
        HealthPollDecision.ProcessExited => MessageProcessExited,
        _ => MessageTimeout,
    };

    public static string IncompatibleDetailMessage(
        string? chunkerVersion,
        string? coreVersion,
        string expectedCoreVersion)
    {
        var engine = $"{NormalizeVersion(coreVersion)}/chunker {chunkerVersion ?? "?"}";
        var expected = $"{NormalizeVersion(expectedCoreVersion)}/chunker {ExpectedChunkerVersion}";
        return $"AI 引擎版本与应用不匹配（引擎 {engine}，应用期望 {expected}）。请更新应用或重启引擎。";
    }

    public static bool ShouldReplaceOrphan(
        string? chunkerVersion,
        string? coreVersion,
        string expectedCoreVersion,
        bool hasBundledSidecar,
        string? orphanExecutable,
        string? bundledExecutable,
        DateTimeOffset? orphanStartedAt,
        DateTimeOffset? bundledModifiedAt)
    {
        if (!IsCompatible(chunkerVersion, coreVersion, expectedCoreVersion))
            return true;
        if (!hasBundledSidecar)
            return LooksLikeFrozenSidecar(orphanExecutable);
        if (!PathsReferToSameFile(orphanExecutable, bundledExecutable))
            return true;
        if (orphanStartedAt is not { } started || bundledModifiedAt is not { } modified)
            return true;
        return started < modified;
    }

    public static bool LooksLikeFrozenSidecar(string? path)
    {
        if (string.IsNullOrEmpty(path)) return false;
        var normalized = path.Replace('\\', '/');
        return normalized.Contains("/lumina-core/lumina-core", StringComparison.OrdinalIgnoreCase)
            || normalized.EndsWith("/lumina-core.exe", StringComparison.OrdinalIgnoreCase);
    }

    public static bool PathsReferToSameFile(string? left, string? right)
    {
        if (string.IsNullOrEmpty(left) || string.IsNullOrEmpty(right)) return false;
        return string.Equals(
            Path.GetFullPath(left),
            Path.GetFullPath(right),
            StringComparison.OrdinalIgnoreCase);
    }

    public static SidecarEngineStatus EngineStatus(
        bool isRunning,
        bool isBootstrapping,
        bool userStopped,
        string? launchError)
    {
        if (isBootstrapping) return SidecarEngineStatus.Starting;
        if (isRunning) return SidecarEngineStatus.Running;
        if (userStopped) return SidecarEngineStatus.Stopped;
        if (!string.IsNullOrEmpty(launchError)) return SidecarEngineStatus.Failed;
        return SidecarEngineStatus.Stopped;
    }

    public static string StatusLabel(SidecarEngineStatus status) => status switch
    {
        SidecarEngineStatus.Starting => "正在启动…",
        SidecarEngineStatus.Running => "运行中",
        SidecarEngineStatus.Failed => "启动失败",
        _ => "已停止",
    };

    public static bool ShouldAutoStart(bool userStopped) => !userStopped;

    public static bool MustKillPortListenerOnStop(bool hasOwnedProcess, bool portOccupied) =>
        hasOwnedProcess || portOccupied;

    public static bool ShouldKillListenerBeforeLaunch(
        bool healthResponded,
        bool shouldReplaceOrphan,
        bool portOccupied)
    {
        if (healthResponded) return shouldReplaceOrphan;
        return portOccupied;
    }

    public static bool ShouldReuseLeftover(bool healthResponded, bool shouldReplaceOrphan) =>
        healthResponded && !shouldReplaceOrphan;
}

public enum SidecarEngineStatus
{
    Starting,
    Running,
    Stopped,
    Failed,
}
